config/base/derive/src/proxy.rs

```diff
@@ -140,30 +116,76 @@ fn impl_load_from_disk(proxy_name: &Ident) -> proc_macro2::TokenStream {
     }
 }
 
-fn gen_proxy_struct_fields(ast: &StructWithFields) -> proc_macro2::TokenStream {
-    let combined_fields = ast.fields.iter().map(|field| {
-        let ident = &field.ident;
-        let ty = &field.ty;
-        quote! {
-            #ident: Option<#ty>
+fn gen_proxy_struct(mut ast: StructWithFields) -> StructWithFields {
+    // Make mut, parse quote and modify ast for proxy
+    ast.fields.iter_mut().for_each(|field| {
+        // For fields of `Configuration` that have an inner config, the corresponding
+        // proxy field should have a proxy there as well
+        if field.has_inner {
+            #[allow(clippy::expect_used)]
+            if let Type::Path(path) = &mut field.ty {
+                let old_ident = &path.path.segments.last().expect("Can't be empty").ident;
+                let new_ident = format_ident!("{}Proxy", old_ident);
+                path.path.segments.last_mut().expect("Can't be empty").ident = new_ident;
+            }
         }
+        let ty = &field.ty;
+        field.ty = parse_quote! {
+            Option<#ty>
+        };
+        field.has_option = true;
+        // Also remove `#[serde(default = ..)]` if present
+        // as it breaks proxy deserialization
+        utils::remove_attr(&mut field.attrs, "serde");
     });
+    ast.parent_ident = Some(ast.ident.clone());
+    ast.ident = format_ident!("{}Proxy", ast.ident);
+    utils::remove_attr_struct(&mut ast, "config");
+    ast
+}
+
+// REVIEW: need lvalue here as well?
+fn impl_build(ast: &StructWithFields) -> proc_macro2::TokenStream {
+    let checked_fields = gen_none_fields_check(ast);
+    let proxy_name = &ast.ident;
+    #[allow(clippy::expect_used)]
+    let parent_name = ast
```

Comment:
I would suggest to add `derive(Builder)` macro with macro attribute like `#[builer(parent = ...)]`, this way we avoid adding new fields to `StructWithFields`.

It should be pretty simple change since `impl_build` already recieves full `StructWithField`.

Reply:
Or instead just do it the old way and pass `impl_build` the parent name together with the struct. Will do that split in the next PR as I was going to anyways.